lib/src/internal/http/http_bucket.dart

```diff
@@ -1,94 +1,74 @@
+  HttpBucket(this._limit, this._remaining, this._reset, this._resetAfter, this._bucketId);
 
-      if (waitTime > 0) {
-        _events.onRateLimitedController.add(RatelimitEvent(request, true));
-        _httpHandler.logger.warning("Rate limited internally on endpoint: ${request.uri}. Trying to send request again in $waitTime ms...");
+  static HttpBucket? fromResponseSafe(http.StreamedResponse response) {
+    int? limit = getLimitFromHeaders(response.headers);
+    int? remaining = getRemainingFromHeaders(response.headers);
+    DateTime? reset = getResetFromHeaders(response.headers);
+    Duration? resetAfter = getResetAfterFromHeaders(response.headers);
+    String? bucketId = getBucketIdFromHeaders(response.headers);
 
-        return Future.delayed(Duration(milliseconds: waitTime), () => execute(request));
-      }
+    if (limit == null || remaining == null || reset == null || resetAfter == null || bucketId == null) {
+      return null;
     }
 
-    // Execute request
-    try {
-      final response = await _httpHandler.httpClient.send(await request.prepareRequest());
+    return HttpBucket(limit, remaining, reset, resetAfter, bucketId);
+  }
 
-      _setBucketValues(response.headers);
-      return response;
-    } on HttpClientException catch (e) {
-      if (e.response == null) {
-        _httpHandler.logger.warning("Http Error on endpoint: ${request.uri}. Error: [${e.message.toString()}].");
-        return Future.error(e);
-      }
+  static String? getBucketIdFromHeaders(Map<String, String> headers) => headers[xRateLimitBucket];
 
-      final response = e.response as http.StreamedResponse;
+  static int? getLimitFromHeaders(Map<String, String> headers) => headers[kRateLimitLimit] == null ? null : int.parse(headers[kRateLimitLimit]!);
 
-      // Check for 429, emmit events and wait given in response body time
-      if (response.statusCode == 429) {
-        final responseBody = jsonDecode(await response.stream.bytesToString());
-        final retryAfter = ((responseBody["retry_after"] as double) * 1000).round();
+  static int? getRemainingFromHeaders(Map<String, String> headers) => headers[kRateLimitRemaining] == null ? null : int.parse(headers[kRateLimitRemaining]!);
 
-        _events.onRateLimitedController.add(RatelimitEvent(request, false, response));
-        _httpHandler.logger.warning("Rate limited via 429 on endpoint: ${request.uri}. Trying to send request again in $retryAfter ms...");
+  // Server-Client clock drift makes headers.reset useless, build reset from headers.resetAfter and DateTime.now()
+  static DateTime? getResetFromHeaders(Map<String, String> headers) =>
+      headers[kRateLimitResetAfter] == null ? null : DateTime.now().add(getResetAfterFromHeaders(headers)!);
 
-        return Future.delayed(Duration(milliseconds: retryAfter), () => execute(request));
-      }
+  static Duration? getResetAfterFromHeaders(Map<String, String> headers) =>
```

Comment:
Static constant?

Reply:
gonna resolve this for now. We are gonna fix that maybe in future if its gonna be a problem